paho/client.go

```diff
@@ -897,15 +898,29 @@ func (c *Client) publishQoS12(ctx context.Context, pb *packets.Publish, o Publis
 	if _, err := pb.WriteTo(c.config.Conn); err != nil {
 		c.debug.Printf("failed to write packet %d to connection: %s", pb.PacketID, err)
 		if o.Method == PublishMethod_AsyncSend {
-			return nil, ErrNetworkErrorAfterStored // Async send, so we don't wait for the response (may add callbacks in the future to enable user to obtain status)
+			if o.OnComplete != nil {
+				go o.OnComplete(nil, fmt.Errorf("failed to write packet to connection: %w", err))
+			}
+			return nil, ErrNetworkErrorAfterStored
 		}
+		return nil, fmt.Errorf("failed to write packet to connection: %w", err)
 	}
 	c.config.PingHandler.PacketSent()
 
 	if o.Method == PublishMethod_AsyncSend {
-		return nil, nil // Async send, so we don't wait for the response (may add callbacks in the future to enable user to obtain status)
+		if o.OnComplete != nil {
+			go func() {
+				pubResp, err := c.waitForResponse(pubCtx, pb, ret)
+				o.OnComplete(pubResp, err)
```

Comment:
This works but will not survive a reconnection so I'm unsure if it offers much benefit. Will add general comments.